src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -193,19 +201,27 @@ public Parser(Compilation compilation, in SourceProductionContext sourceProducti
                         }
                     }
 
+                    INamedTypeSymbol contextTypeSymbol = (INamedTypeSymbol)compilationSemanticModel.GetDeclaredSymbol(classDeclarationSyntax);
+                    Debug.Assert(contextTypeSymbol != null);
+                    string[] diagnosticMessageArgs = new string[] { contextTypeSymbol.Name };
+
                     if (serializableAttributeList == null)
                     {
                         // No types were indicated with [JsonSerializable]
+
+                        // If there are no members on the context, assume that the user wanted source generation and notify them that there are no serializable types.
+                        if (!contextTypeSymbol.MemberNames.Any())
```

Comment:
My feeling is that we should refine this to check for the existence of `GetTypeInfo` overrides. A user that unintentionally adds _any_ member to their context class (be it a helper method, type, etc.) would see this diagnostic disappear completely.